### PR TITLE
add --all-namespaces to export

### PR DESCRIPTION
**Pull request by deads2k**
@bparees @fabianofranz ptal

### CODE REVIEW

File: test/cmd/export.sh [hunk 1]
```diff
@@ -8,6 +8,8 @@ set -o pipefail
 
 oc new-app -f examples/sample-app/application-template-stibuild.json --name=sample
 
+oc export all --all --all-namespaces --exact
```
Comment (bparees) [inline test/cmd/export.sh]: it seems silly that the "all" alias doesn't impicitly also mean "--all"

I guess i could in theory want to say "oc export all things_named_foo" but is that really common?

Comment (deads2k) [inline test/cmd/export.sh]: > it seems silly that the "all" alias doesn't impicitly also mean "--all"
> 
> I guess i could in theory want to say "oc export all things_named_foo" but is that really common?

I'm afraid to look behind the curtain.

File: pkg/cmd/cli/cmd/export.go [hunk 2]
```diff
@@ -99,7 +101,7 @@ func RunExport(f *clientcmd.Factory, exporter Exporter, in io.Reader, out io.Wri
 
 	mapper, typer := f.Object()
 	b := resource.NewBuilder(mapper, typer, f.ClientMapperForCommand()).
-		NamespaceParam(cmdNamespace).DefaultNamespace().
+		NamespaceParam(cmdNamespace).DefaultNamespace().AllNamespaces(allNamespaces).
 		FilenameParam(explicit, filenames...).
 		SelectorParam(selector).
 		ResourceTypeOrNameArgs(all, args...).
```
Comment (deads2k) [inline pkg/cmd/cli/cmd/export.go]: @bparees That would require a change here.  If you passed `true` in place of `all`, you'd get the "normal" behavior you get from `oc get`.  

@smarterclayton what say you?

File: pkg/cmd/cli/cmd/export.go [hunk 3]
```diff
@@ -99,7 +101,7 @@ func RunExport(f *clientcmd.Factory, exporter Exporter, in io.Reader, out io.Wri
 
 	mapper, typer := f.Object()
 	b := resource.NewBuilder(mapper, typer, f.ClientMapperForCommand()).
-		NamespaceParam(cmdNamespace).DefaultNamespace().
+		NamespaceParam(cmdNamespace).DefaultNamespace().AllNamespaces(allNamespaces).
```
Comment (bparees) [inline pkg/cmd/cli/cmd/export.go]: this is magic to me, but if you say that's how it's done, ok......

File: pkg/cmd/cli/cmd/export.go (hunk 2 above)
Comment (bparees) [inline pkg/cmd/cli/cmd/export.go]: i'm going to retract my suggestion.  it's probably worth keeping the consistency w/ other resources even though it's a bit weird.

**Comment by bparees** [comment]
lgtm.

File: pkg/cmd/cli/cmd/export.go (hunk 3 above)
Comment (deads2k) [inline pkg/cmd/cli/cmd/export.go]: > this is magic to me, but if you say that's how it's done, ok......

https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/get.go#L112 as an example.  Basically, its a series of overrides.  "use this namespace", "default the namespace sometimes if needed", "override the use-this-namespace with all".

**Comment by deads2k** [comment]
@bparees second look please.  I removed the `--all` flag.  It is now always `--all`, so it behaves similarly to `oc get`

File: pkg/cmd/cli/cmd/export.go [hunk 4]
```diff
@@ -70,15 +70,18 @@ func NewCmdExport(fullName string, f *clientcmd.Factory, in io.Reader, out io.Wr
 	cmd.Flags().Bool("exact", false, "Preserve fields that may be cluster specific, such as service portalIPs or generated names")
 	cmd.Flags().Bool("raw", false, "If true, do not alter the resources in any way after they are loaded.")
 	cmd.Flags().StringP("selector", "l", "", "Selector (label query) to filter on")
-	cmd.Flags().Bool("all", false, "Select all resources in the namespace of the specified resource types")
+	cmd.Flags().Bool("all-namespaces", false, "If present, list the requested object(s) across all namespaces. Namespace in current context is ignored even if specified with --namespace.")
 	cmd.Flags().VarP(&filenames, "filename", "f", "Filename, directory, or URL to file to use to edit the resource.")
+
+	cmd.Flags().Bool("all", true, "Select all resources in the namespace of the specified resource types")
+	cmd.Flags().MarkDeprecated("all", "all is now ignored")
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/export.go]: explain it is ignored because not specifying a resource name implicitly selects all, right?

**Comment by deads2k** [comment]
[merge]

File: hack/test-end-to-end.sh [hunk 5]
```diff
@@ -72,9 +72,7 @@ function cleanup()
 	oc get -n docker builds --output-version=v1beta3 -t '{{ range .items }}{{.metadata.name}}{{ "\n" }}{{end}}' | xargs -r -l oc build-logs -n docker >"${LOG_DIR}/dockerbuild.log"
 	oc get -n custom builds --output-version=v1beta3 -t '{{ range .items }}{{.metadata.name}}{{ "\n" }}{{end}}' | xargs -r -l oc build-logs -n custom >"${LOG_DIR}/custombuild.log"
 
-	oc export all --all -n test --raw > ${LOG_DIR}/export_all_test.json
-	oc export all --all -n docker --raw > ${LOG_DIR}/export_all_docker.json
-	oc export all --all -n custom --raw > ${LOG_DIR}/export_all_custom.json
+	oc export all --all --all-namespaces --raw -o json > ${LOG_DIR}/export_all.json
```
Comment (bparees) [inline hack/test-end-to-end.sh]: fix this to not use deprecrated flags? :)

**Pull request merged**